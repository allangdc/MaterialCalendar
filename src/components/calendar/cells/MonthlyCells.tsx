import React, { ReactNode, useContext, useEffect, useState } from "react";
import { Grid, Typography } from "@mui/material";
import { isSameDay, isToday } from "date-fns";
import { CalendarContext, CalendarHoliday } from "..";

interface Props {
  id?: string;
  children?: ReactNode;
  day: Date;
}

const MonthlyCells: React.FC<Props> = (props: Props) => {
  const { children, day } = props;
  const { currentDate, holidays, holidayColor } = useContext(CalendarContext);
  const [myHoly, setMyHoly] = useState<CalendarHoliday>();

  useEffect(() => {
    const holy = holidays?.find((dayHoly) => {
      if (isSameDay(dayHoly.day, day)) {
        return true;
      }
      if (dayHoly.yearly) {
        const tempHoly = new Date(dayHoly.day);
        tempHoly.setFullYear(day.getFullYear());
        if (isSameDay(tempHoly, day)) {
          return true;
        }
      }
      return false;
    });
    setMyHoly(holy);
  }, [holidays, day]);

  const colorCells = (): string | undefined => {
    if (isToday(day)) {
      return "#F8AD63";
    }
    if (myHoly) {
      return holidayColor || "#FECEF1";
    }
    if (day.getDay() === 0 || day.getDay() === 6) {
      return "#F7F6F3";
    }
    return undefined;
  };

  const colorText = (): string | undefined => {
    if (day.getMonth() !== currentDate.getMonth()) {
      return "ThreeDHighlight";
    }
    return "gray";
  };

  const HolidayInfo: React.FC = () => {
    return (
      <div style={{ position: "relative" }}>
        <Typography
          align="center"
          variant="subtitle2"
          style={{
            paddingLeft: 10,
            paddingRight: 10,
          }}
        >
          {myHoly && <>{myHoly?.title}</>}
        </Typography>
      </div>
    );
  };

  return (
    <Grid
      container
      data-testid="MonthlyView"
      style={{
        backgroundColor: colorCells(),
        height: "100%",
      }}
    >
      <Grid item xs={12}>
        <Typography
          variant="subtitle2"
          align="right"
          style={{ padding: 5, color: colorText() }}
        >
          {day.getDate()}
        </Typography>
      </Grid>
      <Grid item xs={12}>
        <HolidayInfo />
      </Grid>
      <Grid item xs={12}>
        {children}
      </Grid>
    </Grid>
  );
};

export default MonthlyCells;
