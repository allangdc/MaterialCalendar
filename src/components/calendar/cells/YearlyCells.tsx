import React, { ReactChildren } from "react";
import { Grid } from "@mui/material";
import { isSameMonth, startOfToday } from "date-fns";

interface Props {
  id?: string;
  children?: ReactChildren;
  day: Date;
}

const YearlyCells: React.FC<Props> = (props: Props) => {
  const { children, day } = props;
  const currentDate = startOfToday();

  const colorCells = (): string | undefined => {
    if (isSameMonth(day, currentDate)) {
      return "#F8AD63";
    }
    return undefined;
  };

  return (
    <Grid container style={{ backgroundColor: colorCells(), minHeight: 50 }}>
      <Grid item xs={12}>
        {children}
      </Grid>
    </Grid>
  );
};

export default YearlyCells;
