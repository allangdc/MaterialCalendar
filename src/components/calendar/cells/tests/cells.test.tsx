import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import { hexToRgb } from "@mui/system";
import MonthlyCells from "../MonthlyCells";
import YearlyCells from "../YearlyCells";
import Cells from "..";
import { CalendarContext, CalendarFormat, ICalendarContext } from "../..";
import {
  addDays,
  endOfMonth,
  startOfMonth,
  startOfToday,
  subDays,
} from "date-fns";
import renderer from "react-test-renderer";

const dateRef = startOfToday();

const dummyValue: ICalendarContext = {
  formatCal: CalendarFormat.MONTHLY,
  currDate: dateRef,
  setCurrDate: function (): void {
    throw new Error("Function not implemented.");
  },
  width: undefined,
  setFormatCal: function (): void {
    throw new Error("Function not implemented.");
  },
  language: "",
  holidays: undefined,
  holidayColor: undefined,
  lineColor: undefined,
  dataArray: [],
  setDataArray: function (): void {
    throw new Error("Function not implemented.");
  },
  headerArray: [],
  setHeaderArray: function (): void {
    throw new Error("Function not implemented.");
  },
  sameWidthP100: undefined,
};

test("Must shows YearlyCells", () => {
  dummyValue.formatCal = CalendarFormat.YEARLY;
  render(
    <CalendarContext.Provider value={dummyValue}>
      <Cells day={dateRef} />
    </CalendarContext.Provider>
  );
  const linkElement = screen.queryByText(dateRef.getDate());
  expect(linkElement).not.toBeInTheDocument();
});

test("Yearly cells cannot show days", () => {
  render(<YearlyCells day={dateRef} />);
  const linkElement = screen.queryByText(dateRef.getDate());
  expect(linkElement).not.toBeInTheDocument();
});

test("Check days in cells without Children", () => {
  dummyValue.formatCal = CalendarFormat.MONTHLY;
  render(
    <CalendarContext.Provider value={dummyValue}>
      <Cells day={dateRef} />
    </CalendarContext.Provider>
  );
  const linkElement = screen.getByText(dateRef.getDate());
  expect(linkElement).toBeInTheDocument();
});

test("Check days in cells with Children", () => {
  render(
    <CalendarContext.Provider value={dummyValue}>
      <Cells day={dateRef}>
        <h1>Hello@World</h1>
      </Cells>
    </CalendarContext.Provider>
  );
  const linkDay = screen.getByText(dateRef.getDate());
  const linkChildren = screen.getByText("Hello@World");
  expect(linkDay).toBeInTheDocument();
  expect(linkChildren).toBeInTheDocument();
});

test("Background color for weekend days should be gray for Monthly view", () => {
  const dates = [
    new Date(2021, 0, 2), // sat
    new Date(2021, 0, 3), // sun
  ];
  dates.map((dt) => {
    cleanup();
    render(
      <CalendarContext.Provider value={dummyValue}>
        <Cells day={dt} />
      </CalendarContext.Provider>
    );
    const element = screen.getByTestId("MonthlyView");
    const styles = getComputedStyle(element);
    expect(styles.backgroundColor).toBe(hexToRgb("#F7F6F3"));
  });
});

test("Background color for normal days should be undefined", () => {
  const dates = [
    new Date(2021, 0, 4), // mon
    new Date(2021, 0, 5), // tue
    new Date(2021, 0, 6), // wed
    new Date(2021, 0, 7), // thu
    new Date(2021, 0, 8), // fri
  ];
  dummyValue.formatCal = CalendarFormat.YEARLY;
  dates.map((dt) => {
    cleanup();
    render(
      <CalendarContext.Provider value={dummyValue}>
        <Cells day={dt} />
      </CalendarContext.Provider>
    );
    const element = screen.getByTestId("YearlyView");
    const styles = getComputedStyle(element);
    expect(styles.backgroundColor).toBe("");
  });

  dummyValue.formatCal = CalendarFormat.MONTHLY;
  dates.map((dt) => {
    cleanup();
    render(
      <CalendarContext.Provider value={dummyValue}>
        <Cells day={dt} />
      </CalendarContext.Provider>
    );
    const element = screen.getByTestId("MonthlyView");
    const styles = getComputedStyle(element);
    expect(styles.backgroundColor).toBe("");
  });
});

test("Background color for today should be orange in monthly view", () => {
  const dt = dateRef; // today
  render(<MonthlyCells day={dt} />);
  const element = screen.getByTestId("MonthlyView");
  const styles = getComputedStyle(element);
  expect(styles.backgroundColor).toBe(hexToRgb("#F8AD63"));
});

test("Background color for current month should be orange in yearly view", () => {
  const dt = dateRef; // today
  dummyValue.formatCal = CalendarFormat.YEARLY;
  render(
    <CalendarContext.Provider value={dummyValue}>
      <Cells day={dt} />
    </CalendarContext.Provider>
  );
  const element = screen.getByTestId("YearlyView");
  const styles = getComputedStyle(element);
  expect(styles.backgroundColor).toBe(hexToRgb("#F8AD63"));
});

test("Textcolor for days in current month must be gray in Monthly view", () => {
  dummyValue.formatCal = CalendarFormat.MONTHLY;
  const dtinit = startOfMonth(dateRef); // same month of today
  const dtend = endOfMonth(dateRef); // same month of today
  const dates = [dtinit, dtend];
  dates.map((dt) => {
    cleanup();
    render(
      <CalendarContext.Provider value={dummyValue}>
        <Cells day={dt} />
      </CalendarContext.Provider>
    );
    const element = screen.queryByText(dt.getDate());
    expect(element).toBeInTheDocument();
    if (element) {
      const styles = getComputedStyle(element);
      expect(styles.color).toBe("gray");
    }
  });
});

test("Textcolor for days out current month must be lightgray in Monthly view", () => {
  dummyValue.formatCal = CalendarFormat.MONTHLY;
  const dtinit = subDays(startOfMonth(dateRef), 1); // before month of today
  const dtend = addDays(endOfMonth(dateRef), 1); // after same month of today
  const dates = [dtinit, dtend];
  dates.map((dt) => {
    cleanup();
    render(
      <CalendarContext.Provider value={dummyValue}>
        <Cells day={dt} />
      </CalendarContext.Provider>
    );
    const element = screen.queryByText(dt.getDate());
    expect(element).toBeInTheDocument();
    if (element) {
      const styles = getComputedStyle(element);
      expect(styles.color).toBe("ThreeDHighlight");
    }
  });
});

test("Day must be align on the right side of the cell in monthly view", () => {
  const dt = dateRef; // today
  dummyValue.formatCal = CalendarFormat.MONTHLY;
  render(
    <CalendarContext.Provider value={dummyValue}>
      <Cells day={dt} />
    </CalendarContext.Provider>
  );
  const element = screen.getByText(dt.getDate());
  expect(element).toBeInTheDocument();
  if (element) {
    const styles = getComputedStyle(element);
    expect(styles.textAlign).toBe("right");
  }
});

test("Snapshot of the cells in monthly view", () => {
  dummyValue.formatCal = CalendarFormat.MONTHLY;
  const myDate = new Date(1984, 5, 13);
  const component = renderer.create(
    <CalendarContext.Provider value={dummyValue}>
      <Cells day={myDate} />
    </CalendarContext.Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Snapshot of the cells in yearly view", () => {
  dummyValue.formatCal = CalendarFormat.YEARLY;
  const myDate = new Date(1984, 5, 13);
  const component = renderer.create(
    <CalendarContext.Provider value={dummyValue}>
      <Cells day={myDate} />
    </CalendarContext.Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
