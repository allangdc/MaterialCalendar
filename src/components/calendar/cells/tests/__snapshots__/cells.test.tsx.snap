// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of the cells in monthly view 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
    data-testid="MonthlyView"
    style={
      Object {
        "backgroundColor": undefined,
        "minHeight": 50,
      }
    }
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 MuiTypography-alignRight css-h2l1uj-MuiTypography-root"
        style={
          Object {
            "color": "ThreeDHighlight",
            "padding": 5,
          }
        }
      >
        13
      </h6>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
    />
  </div>
</div>
`;

exports[`Snapshot of the cells in yearly view 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
    data-testid="YearlyView"
    style={
      Object {
        "backgroundColor": undefined,
        "minHeight": 50,
      }
    }
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
    />
  </div>
</div>
`;
